Docs for External Sessions Data store

krmahadevan opened the pull request:
**Thanks for contributing to the Selenium site and documentation!**
**A PR well described will help maintainers to review and merge it quickly**

Before submitting your PR, please check our [contributing](https://selenium.dev/documentation/en/contributing/) guidelines.
Avoid large PRs, and help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

Adding documentation to explain how to use an external data store for storing sessions.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

The documentation for this is right now not there, even though the implementation already exists.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Change to the site (I have double-checked the Netlify deployment, and my changes look good)
- [X] Code example added (and I also added the example to all translated languages)
- [ ] Improved translation
- [ ] Added new translation (and I also added a notice to each document missing translation)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] I have read the [**contributing**](https://selenium.dev/documentation/en/contributing/) document.
- [X] I have used [hugo](https://gohugo.io/) to render the site/docs locally and I am sure it works.
<!--- Provide a general summary of your changes in the Title above -->

krmahadevan committed:
Docs for External Sessions Data store

krmahadevan commented:
@diemol - Please let me know if there's anything that needs to be changed in this PR.

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+---
+title: "External datastore for sessions"
+linkTitle: "External datastore for sessions"
```
```suggestion
title: "External datastore"
linkTitle: "External datastore"
```

diemol replied:
Let's leave it "External datastore" only, as the goal in the future is to have this also for the Distributor. So both things could go here. Also the file name could change from `external_datastore_sessions.en.md` to `external_datastore.en.md`

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+* [Redis backed Session Map](#redis-backed-session-map) 
+    * [Steps](#steps)
+
+## Why an external datastore ?
+
+When you are running a distributed grid that comprises of:
+
+* Event Bus
+* New Session Queue
+* Session Map
+* Distributor
+* Router
+* Node
+
+you may want to scale these individual components horizontally (i.e., run more than one instances of the same component. For e.g., two `Session Map`) to cater to your grid's increased load.
+
+What would happen if we run two instances of a `Session Map`? 
+
+By default, a `Session Map` component uses an in-memory data store to store sessions information (which session is running on which node).
+
+So if there are two `Session Map` instances, then we end up with 2 in-memory data store instances which are not shared.
+
+This can cause a lot of issues such as:
+
+* Sessions data not being shared by both the instances
+* Session data query goes to the wrong instance of `Session Map`
+* Sessions data is lost if one of the instances goes down
+
+In short we now have a consistency problem that is so closely associated with the notion of microservices.
```
With the current Grid 4, this is not accurate because it is not possible to run two `SessionMap` instances. Well, it is technically possible, but the `Router` can only talk to only one `SessionMap` instance.

The reasoning behind backing the data from the `SessionMap` is to allow the user to take down the `SessionMap` instance and bring a new one without loosing the current data.

As you can see, the topic of a high availability Grid has lots of room for improvement.

krmahadevan replied:
Wouldn't it work fine if we put up a reverse proxy/load balancer (maybe something like a haproxy or nginx) which just does routing on a round robin basis? In that case, the router would just connect to the load balancer that fronts the session map and so it should be possible no ?

diemol replied:
Yes, but that is out of the scope of this doc for now.

That is what I tried to say with
>As you can see, the topic of a high availability Grid has lots of room for improvement.

We need to figure out ways for users to have a high availability Grid. What you propose is an alternative.

krmahadevan replied:
> We need to figure out ways for users to have a high availability Grid. 

I feel that if Grid helps users to leverage High Availability Tools (which are meant for that purpose) then we wouldn't need to build them (I hope that we aren't looking at that) and so the ask of high availability would have been resolved right ?

diemol replied:
Yes, we should reuse existing tools as much as possible. But as the current state, we can only document what we know that works. 

A future step would be to figure out what tools are needed and how to combine them with Grid to make it highly available. 

We can continue this conversation in the #selenium-tlc channel on Slack 🙏

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+
+## The solution
+
+The simplest solution would be to ensure that all the sessions map instances use a common shared data store. And thus we have a need to externalise this data store which holds running sessions information.
```
This needs to be reworked based on the previous comment.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+
+The simplest solution would be to ensure that all the sessions map instances use a common shared data store. And thus we have a need to externalise this data store which holds running sessions information.
+
+Thanksfully Grid4 is already geared up to support this use case. 
```
```suggestion
Thankfully Grid already supports this use case. 
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+The simplest solution would be to ensure that all the sessions map instances use a common shared data store. And thus we have a need to externalise this data store which holds running sessions information.
+
+Thanksfully Grid4 is already geared up to support this use case. 
+You can use a data store that is: 
+
+* Backed by your favourite database (or) 
+* Backed by a Redis Cache system
+
+so that all the running sessions map instances use a single shared data store.
```
```suggestion
You can use a data store that is backed by your favorite database or by a Redis cache system.
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+docker-compose up -d
+```
+
+_So our database name is `selenium_sessions` with its username and password set to `seluser`_
````
```suggestion
_Our database name is `selenium_sessions` with its username and password set to `seluser`_
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+
+_Note:_ If you plan to use an existing PostGreSQL DB instance, then replace `localhost:5432` with the actual host and port number of your instance.
+
+* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed grid.
```
```suggestion
* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed Grid.
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed grid.
+
+```bash
+SE_VERSION=4.6.0
````
```suggestion
SE_VERSION=<current_selenium_version>
```

diemol replied:
I prefer to leave this placeholder because it makes the docs easier to maintain.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting New Session Queue'
+java -jar $JAR_NAME sessionqueue --port 5559 &
+#Start session map
+echo 'Starting sessions manager'
```
```suggestion
echo 'Starting Session Map'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-jdbc:${SE_VERSION} org.postgresql:postgresql:42.3.1) \
+sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &
+#Start distributor
+echo 'Starting distributor'
```
```suggestion
echo 'Starting Distributor'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+SUBSCRIBE="--subscribe-events tcp://localhost:4443"
+SESSIONS="--sessions http://localhost:5556"
+SESSIONS_QUEUE="--sessionqueue http://localhost:5559"
+# Start event bus
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+# Start event bus
+echo 'Starting Event Bus'
+java -jar $JAR_NAME event-bus $PUBLISH $SUBSCRIBE --port 5557 &
+#Start New Session Queue
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+java -jar $JAR_NAME \
+--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-jdbc:${SE_VERSION} org.postgresql:postgresql:42.3.1) \
+sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &
+#Start distributor
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+#Start distributor
+echo 'Starting distributor'
+java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &
+#Start router
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting distributor'
+java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &
+#Start router
+echo 'Starting router'
```
```suggestion
echo 'Starting Router'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+#Start router
+echo 'Starting router'
+java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &
+#Start node
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting router'
+java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &
+#Start node
+echo 'Starting node'
```
```suggestion
echo 'Starting Node'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+    * `sessions.toml`
+    * `distributed.sh`
+
+* You can now spawn the grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores (inserts and then deletes) session information into the PostGreSQL database.
```
```suggestion
* You can now spawn the Grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores session information into the PostGreSQL database.
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+
+* You can now spawn the grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores (inserts and then deletes) session information into the PostGreSQL database.
+
+In the line which spawns a `Session Map` on a machine:
```
```suggestion
In the line which spawns a `SessionMap` on a machine:
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+      - "6379:6379"
+```
+
+We can now start our redis container by running:
````
```suggestion
We can now start our Redis container by running:
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+docker-compose up -d
+```
+
+* Create a Selenium configuration file named `sessions.toml` with the below contents:
````
```suggestion
* Create a Selenium Grid configuration file named `sessions.toml` with the below contents:
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+SUBSCRIBE="--subscribe-events tcp://localhost:4443"
+SESSIONS="--sessions http://localhost:5556"
+SESSIONS_QUEUE="--sessionqueue http://localhost:5559"
+# Start event bus
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed grid.
+
+```bash
+SE_VERSION=4.6.0
````
```suggestion
SE_VERSION=<current_selenium_session>
```

diemol replied:
I prefer to have a placeholder to simplify docs maintenance.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+# Start event bus
+echo 'Starting Event Bus'
+java -jar $JAR_NAME event-bus $PUBLISH $SUBSCRIBE --port 5557 &
+#Start New Session Queue
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+#Start New Session Queue
+echo 'Starting New Session Queue'
+java -jar $JAR_NAME sessionqueue --port 5559 &
+#Start session map
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting New Session Queue'
+java -jar $JAR_NAME sessionqueue --port 5559 &
+#Start session map
+echo 'Starting sessions manager'
```
```suggestion
echo 'Starting Session Map'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+java -jar $JAR_NAME \
+--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-redis:${SE_VERSION}) \
+sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &
+#Start distributor
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-redis:${SE_VERSION}) \
+sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &
+#Start distributor
+echo 'Starting distributor'
```
```suggestion
echo 'Starting Distributor'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+#Start distributor
+echo 'Starting distributor'
+java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &
+#Start router
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting distributor'
+java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &
+#Start router
+echo 'Starting router'
```
```suggestion
echo 'Starting Router'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+#Start router
+echo 'Starting router'
+java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &
+#Start node
```
```suggestion
```

diemol replied:
No need to write this as the `echo` from the next line already says the same.

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+echo 'Starting router'
+java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &
+#Start node
+echo 'Starting node'
```
```suggestion
echo 'Starting Node'
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
```diff
@@ -0,0 +1,272 @@
+    * `sessions.toml`
+    * `distributed.sh`
+
+* You can now spawn the grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores (inserts and then deletes) session information into the Redis instance. You can perhaps make use of a Redis GUI such as [TablePlus](https://tableplus.com/) to see them (Make sure that you have setup a debug point in your test, because the values will get deleted a ...
```
```suggestion
* You can now spawn the Grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores session information into the Redis instance. You can use a Redis GUI such as [TablePlus](https://tableplus.com/) to see them (make sure that you have setup a debug point in your test, because the values will get deleted as soon as the test runs to completion).
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+In the line which spawns a `Session Map` on a machine:
+
+```bash
+export SE_VERSION=4.6.0
````
```suggestion
export SE_VERSION=<current_selenium_version>
```

krmahadevan replied:
Fixed

diemol commented on website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md:
````diff
@@ -0,0 +1,272 @@
+In the line which spawns a `Session Map` on a machine:
+
+```bash
+export SE_VERSION=4.6.0
````
```suggestion
export SE_VERSION=<current_selenium_version>
```

krmahadevan replied:
Fixed

krmahadevan commented:
@diemol - I have addressed all your review comments, amended the commit with the changes and also rebased off of `trunk`. Please help take a look when you get the time.

diemol reviewed (approved):
Thank you, @krmahadevan!
